contracts/SimpleGAMultiSig.aes

```diff
@@ -0,0 +1,153 @@
+          true => set_current_tx(ga_tx_hash, ttl)
+  
+  // signer confirms tx
+  stateful entrypoint confirm(sig : signature, n : int) = 
+    check_nonce(n)
+    switch(state.current_tx)
+      None => abort("ERROR_NOTHING_TO_CONFIRM")
+      Some(tx) =>
+        require(is_authorized_and_valid(Call.caller, tx.ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+        require(!List.contains(Call.caller, tx.confirmed_by),"ERROR_ALREADY_CONFIRMED")
+        require(!is_expired(), "ERROR_TX_ALREADY_EXPIRED")        
+        let updated_tx = tx{confirmed_by=Call.caller :: tx.confirmed_by}
+        put(state{current_tx=Some(updated_tx)})
+        Chain.event(TxConfirmed(tx.ga_tx_hash, Call.caller))
+        if(has_consensus())
+          Chain.event(TxConsensusReached(tx.ga_tx_hash))
+  
+  // revoke the current tx and clean state
+  stateful entrypoint revoke(sig : signature, n: int) =
+    check_nonce(n)
+    switch(state.current_tx)
+      None => abort("ERROR_NOTHING_TO_REVOKE")
+      Some(tx) =>
+        require(is_authorized_and_valid(Call.caller, tx.ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+        put(state{current_tx = None})
+        Chain.event(TxRevoked(Call.caller))
+
+  // check if given address is within the list of multisig_users
+  // AND 
+  // check that the signature of the proposed ga tx was made using the private key of the given address
+  entrypoint is_authorized_and_valid(addr : address, ga_tx_hash_to_verify : hash, sig : signature) : bool =
+    List.contains(addr, state.multisig_users) && Crypto.verify_sig(ga_tx_hash_to_verify, addr, sig)
+
+  // validate if enough signers have confirmed
+  entrypoint has_consensus() : bool = 
+    switch(state.current_tx)
+      None => false
+      Some(tx) => List.length(tx.confirmed_by) >= state.confirmations_required
+	
+  // check if current_tx is already expired
+  function is_expired() : bool =
+    switch(state.current_tx)
+      None => true
+      Some(tx) => Chain.block_height > tx.expires_at
+
+  // set current tx to state and emit new proposed event
+  stateful function set_current_tx(p_ga_tx_hash : hash, ttl : Chain.ttl) =
+    let new_tx = { ga_tx_hash = p_ga_tx_hash,
+                   expires_at = expiration_height(ttl),
+                   proposed_by = Call.caller,
+                   confirmed_by = [Call.caller] } 
+    put(state{current_tx = Some(new_tx)})
+    Chain.event(TxProposed(new_tx.ga_tx_hash, new_tx.proposed_by, new_tx.expires_at))
+
+  stateful function check_nonce(n : int) = 
+    require(n >= state.nonce, "ERROR_NONCE_TOO_LOW")
+    require(n =< state.nonce, "ERROR_NONCE_TOO_HIGH")
+    put(state{ nonce = n + 1 })
+
+  // compute the expiration height depending on the ttl type
+  function expiration_height(ttl : Chain.ttl) : int =
+    switch(ttl)
+      FixedTTL(height) => height
+      RelativeTTL(relative_ttl) => Chain.block_height + relative_ttl
+
+  // return contracts current nonce
+  entrypoint get_nonce() : int =
```

Comment:
add more getters if needed?

Reply:
yes, we at least need to know all the signers

Reply:
added following getters:

- `get_signers() : list(address)`
- `get_consensus_info() : consensus_info`

where `consensus_info` is defined as follows:

```sophia
  record consensus_info =
    { ga_tx_hash : option(hash),
      confirmations_required : int,
      confirmed_by : list(address),
      can_authorize : bool }
```